### ISSUE TITLE
twisted.mail.relaymanager.MXCalculator does not follow CNAMEs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2953]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2953|
|Type|defect|
|Created|2007-12-13 02:29:07Z|
|Branch|https://github.com/twisted/twisted/tree/mx-cname-2953-2|

RFC 974, the second full paragraph on page three:

```
   There is one other special case.  If the response contains an answer
   which is a CNAME RR, it indicates that REMOTE is actually an alias
   for some other domain name. The query should be repeated with the
   canonical domain name.
```

If ```MXCalculator._cbMX``` receives a ```Record_CNAME``` instead of a ```Record_MX``` it will probably fail with an ```AttributeError``` when it tries to get the ```preference```, something CNAMEs don't have.  Even if it could get past this, it would need to do another MX lookup for the name the CNAME refers to.

<details><summary>Searchable metadata</summary>

```
trac-id__2953 2953
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_mx_cname_2953_2 branches/mx-cname-2953-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1197512947000000 1197512947000000
changetime__1200836113000000 1200836113000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [22075]) Branching to 'mx-cname-2953'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

ready for review

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

 * not totally related, but test_mail is huge, it would be great to split it in smaller test files
 * in _cbMx, the following lines are not tested:
```
                t = time.time() - self.badMXs[host]
                if t > 0:
                    del self.badMXs[host]
                    return record

```
 * in _cbMx, the line ```return exchanges[0]``` is not tested either
 * in _cbMx, the exchanges list used to be sorted. There is no test failing, but that may just indicate that the coverage is poor. Are you sure that the sort is useless?
 * in _cbMx, seenAliases ought to be a set.

Except the sort problem, everything is moderately related to your changes, so not mandatory for the merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [22357]) Branching to 'mx-cname-2953-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Thanks for an awesome review.

Hopefully someday test_mail will be split up.  I don't think I'll try to tackle that now. :)

I added tests for the three untested things you pointed out and changed ```seenAliases``` to a set.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [22370]) Merge mx-cname-2953-2

Author: exarkun
Reviewer: therve
Fixes [#2953](https://github.com/twisted/twisted/issues/2953)

Add support to twisted.mail.relaymanager.MXCalculator for recognizing CNAME
records and resolving them, if possible.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
